lib/json_schemer/schema/base.rb

```diff
@@ -29,29 +30,39 @@ def merge(
         :eol => '\z'
       }.freeze
 
+      INSERT_DEFAULT_PROPERTY = proc do |data, property, property_schema|
+        if !data.key?(property) && property_schema.is_a?(Hash) && property_schema.key?('default')
+          data[property] = property_schema.fetch('default').clone
+        end
+      end
+
       def initialize(
         schema,
         format: true,
         insert_property_defaults: false,
+        before_property_validation: nil,
+        after_property_validation: nil,
         formats: nil,
         keywords: nil,
         ref_resolver: DEFAULT_REF_RESOLVER
       )
         raise InvalidSymbolKey, 'schemas must use string keys' if schema.is_a?(Hash) && !schema.empty? && !schema.first.first.is_a?(String)
         @root = schema
         @format = format
-        @insert_property_defaults = insert_property_defaults
+        @before_property_validation = Array(before_property_validation)
+        @before_property_validation.unshift(INSERT_DEFAULT_PROPERTY) if insert_property_defaults
```

Comment:
I didn't realize this before, but I think this will modify the input array:

```
>> x = []
>> y = Array(x)
>> x.object_id == y.object_id
=> true
>> y.unshift(1)
=> [1]
>> x
=> [1]
```

I think splatting the value works similarly and generates a new array:

```
>> x = []
>> y = [*x]
>> x.object_id == y.object_id
=> false
>> y.unshift(1)
=> [1]
>> x
=> []
>> y
=> [1]
```

Reply:
Changed this a [test about not modifying the array](https://github.com/davishmcclurg/json_schemer/pull/77/commits/8d1f8353eec6cc333d75c4ca0a28cefc3e3c1466#diff-05e8376a74a8d91443c8d001e944703cR354)